README.md
What is bukapilot?
------
***Todo: Add some images

[bukapilot](http://github.com/kommuai/bukapilot) is an open source software for advanced driver's assistant system (ADAS). It is a maintained patchset based on [openpilot](http://github.com/commaai/openpilot) which aims to support all Malaysian vehicles especially for Perodua and Proton by using hardwares specially designed for them, while adding complete right hand drive support on top of the existing capability from openpilot.

bukapilot performs the functions of Adaptive Cruise Control (ACC), Lane Keep Assist (LKA), Forward Collision Warning (FCW) and Lane Departure Warning (LDW) for a growing variety of supported [car makes, models and model years](#supported-cars). In addition, while bukapilot is engaged, a camera based Driver Monitoring (DM) feature alerts distracted and asleep drivers.

Integration with Stock Features
------
All supported vehicles:
* Stock LKA is replaced by bukapilot LKA, which only functions when bukapilot is engaged by the user.
* Stock LDW is replaced by bukapilot LDW.

Additionally, on specific supported cars (see ACC column in [supported cars](#supported-cars)):
* Stock ACC is replaced by bukapilot ACC.
* bukapilot FCW operates in addition to stock FCW.

bukapilot should preserve all other vehicle's stock features, including, but are not limited to: FCW, Automatic Emergency Braking (AEB), auto high-beam, blind spot warning, and side collision warning.

Supported Cars
------

Bukapilot patchset modifies the code heavily to integrate with our hardware and backend services, thus we can't make any statement for openpilot supported cars.  Instead, we plan to offer commerical support for the following:

| Make      | Model                         | ADAS     | LKAS                      | Gas                        | Brake         |
| ----------| ------------------------------| ---------| --------------------------| ---------------------------| --------------|
| Perodua   | Axia Advanced 2019            | ASA2.0   | KommuActuator<sup>1</sup> | KommuActuator<sup>1</sup>  | None          |

<sup>1</sup>KommuActuator is a hardware used to provide steering control and stop-and-go capability to some of the bukapilot-supported cars that don't currently support CAN-enabled controls.

KommuAssist Installation Instructions
------
Install bukapilot on a supported device by entering ``https://s4.kommu.ml/upload/bp/installer_openpilot`` during the installer setup.

Follow these [video instructions](https://youtu.be/FktYvHD1PD0) to properly mount the device on the windshield. Note: bukapilot features an automatic pose calibration routine and bukapilot performance should not be affected by small pitch and yaw misalignments caused by imprecise device mounting.

You will be able to engage bukapilot after reviewing the onboarding screens and finishing the calibration procedure.

Limitations of bukapilot LKA and LDW
------

bukapilot LKA and bukapilot LDW do not automatically drive the vehicle or reduce the amount of attention that must be paid to operate your vehicle. The driver must always keep control of the steering wheel and be ready to correct the bukapilot LKA action at all times.

While changing lanes, bukapilot is not capable of looking next to you or checking your blind spot. Only nudge the wheel to initiate a lane change after you have confirmed it's safe to do so.

Many factors can impact the performance of bukapilot LKA and bukapilot LDW, causing them to be unable to function as intended. These include, but are not limited to:

* Poor visibility (heavy rain, snow, fog, etc.) or weather conditions that may interfere with sensor operation.
* The road facing camera is obstructed, covered or damaged by mud, ice, snow, etc.
* Obstruction caused by applying excessive paint or adhesive products (such as wraps, stickers, rubber coating, etc.) onto the vehicle.
* The device is mounted incorrectly.
* When in sharp curves, like on-off ramps, intersections etc...; bukapilot is designed to be limited in the amount of steering torque it can produce.
* In the presence of restricted lanes or construction zones.
* When driving on highly banked roads or in presence of strong cross-wind.
* Extremely hot or cold temperatures.
* Bright light (due to oncoming headlights, direct sunlight, etc.).
* Driving on hills, narrow, or winding roads.

The list above does not represent an exhaustive list of situations that may interfere with proper operation of bukapilot components. It is the driver's responsibility to be in control of the vehicle at all times.

Limitations of bukapilot ACC and FCW
------

bukapilot ACC and bukapilot FCW are not systems that allow careless or inattentive driving. It is still necessary for the driver to pay close attention to the vehicle’s surroundings and to be ready to re-take control of the gas and the brake at all times.

Many factors can impact the performance of bukapilot ACC and bukapilot FCW, causing them to be unable to function as intended. These include, but are not limited to:

* Poor visibility (heavy rain, snow, fog, etc.) or weather conditions that may interfere with sensor operation.
* The road facing camera or radar are obstructed, covered, or damaged by mud, ice, snow, etc.
* Obstruction caused by applying excessive paint or adhesive products (such as wraps, stickers, rubber coating, etc.) onto the vehicle.
* The device is mounted incorrectly.
* Approaching a toll booth, a bridge or a large metal plate.
* When driving on roads with pedestrians, cyclists, etc...
* In presence of traffic signs or stop lights, which are not detected by bukapilot at this time.
* When the posted speed limit is below the user selected set speed. bukapilot does not detect speed limits at this time.
* In presence of vehicles in the same lane that are not moving.
* When abrupt braking maneuvers are required. bukapilot is designed to be limited in the amount of deceleration and acceleration that it can produce.
* When surrounding vehicles perform close cut-ins from neighbor lanes.
* Driving on hills, narrow, or winding roads.
* Extremely hot or cold temperatures.
* Bright light (due to oncoming headlights, direct sunlight, etc.).
* Interference from other equipment that generates radar waves.

The list above does not represent an exhaustive list of situations that may interfere with proper operation of bukapilot components. It is the driver's responsibility to be in control of the vehicle at all times.

Limitations of bukapilot DM
------

bukapilot DM should not be considered an exact measurement of the alertness of the driver.

Many factors can impact the performance of bukapilot DM, causing it to be unable to function as intended. These include, but are not limited to:

* Low light conditions, such as driving at night or in dark tunnels.
* Bright light (due to oncoming headlights, direct sunlight, etc.).
* The driver's face is partially or completely outside field of view of the driver facing camera.
* The driver facing camera is obstructed, covered, or damaged.

The list above does not represent an exhaustive list of situations that may interfere with proper operation of bukapilot components. A driver should not rely on bukapilot DM to assess their level of attention.

User Data and Kommu Account
------

By default, bukapilot uploads the driving data to our servers. You can also access your data by pairing with the KommuApp ([iOS](https://apps.apple.com/us/app/comma-connect/id1456551889), [Android](https://play.google.com/store/apps/details?id=ai.comma.connect&hl=en_US)). We use your data to train better models and improve bukapilot for everyone.

bukapilot logs the road facing camera, CAN, GPS, IMU, magnetometer, thermal sensors, crashes, and operating system logs.
The driver facing camera is logged by default but you can explicitly opt-out in settings. The microphone is not recorded.

By using bukapilot, you agree to [our Privacy Policy](https://kommu.ai/privacy/). You understand that use of this software or its related services will generate certain types of user data, which may be logged and stored at the sole discretion of Kommu. By accepting this agreement, you grant an irrevocable, perpetual, worldwide right to Kommu for the use of this data.

Safety and Testing
----
DISCLAIMER: bukapilot is still under development by a limited number of members. Duty of care will be taken to ensure the safety of both the hardware and the software. 

As a patchset of openpilot, bukapilot benefits from the software safety model of openpilot.  Incremental tests are being implemented to ensure the correctness of the patchset on top of the openpilot release.

Community and Contributing
------

bukapilot is developed by a local Malaysian team [Kommu](https://kommu.ai/) and by users like you. We welcome both pull requests and issues on [GitHub](http://github.com/kommuai/bukapilot). Bug fixes and new car ports are encouraged.

And [follow us on Instagram](https://www.instagram.com/kommu.ai/).

Directory Structure
------
    .
    ├── cereal              # The messaging spec and libs used for all logs
    ├── common              # Library like functionality we've developed here
    ├── installer/updater   # Manages auto-updates of bukapilot
    ├── opendbc             # Files showing how to interpret data from cars
    ├── panda               # Code used to communicate on CAN
    ├── phonelibs           # External libraries
    ├── pyextra             # Extra python packages not shipped in NEOS
    └── selfdrive           # Code needed to drive the car
        ├── assets          # Fonts, images, and sounds for UI
        ├── athena          # Allows communication with the app
        ├── boardd          # Daemon to talk to the board
        ├── camerad         # Driver to capture images from the camera sensors
        ├── car             # Car specific code to read states and control actuators
        ├── common          # Shared C/C++ code for the daemons
        ├── controls        # Planning and controls
        ├── debug           # Tools to help you debug and do car ports
        ├── locationd       # Precise localization and vehicle parameter estimation
        ├── logcatd         # Android logcat as a service
        ├── loggerd         # Logger and uploader of car data
        ├── modeld          # Driving and monitoring model runners
        ├── proclogd        # Logs information from proc
        ├── sensord         # IMU interface code
        ├── test            # Unit tests, system tests, and a car simulator
        └── ui              # The UI

Licensing
------

bukapilot is released under the MIT license. Some parts of the software are released under other licenses as specified.

Any user of this software shall indemnify and hold harmless against Kommu Sdn Bhd. and its directors, officers, employees, agents, stockholders, affiliates, subcontractors and customers from and against all allegations, claims, actions, suits, demands, damages, liabilities, obligations, losses, settlements, judgments, costs and expenses (including without limitation attorneys’ fees and costs) which arise out of, relate to or result from any use of this software by user.

**THIS IS ALPHA QUALITY SOFTWARE FOR RESEARCH PURPOSES ONLY. THIS IS NOT A PRODUCT.
YOU ARE RESPONSIBLE FOR COMPLYING WITH LOCAL LAWS AND REGULATIONS.
NO WARRANTY EXPRESSED OR IMPLIED.**

---

</img> <img src="https://cdn-images-1.medium.com/max/1600/1*C87EjxGeMPrkTuVRVWVg4w.png" width="225"></img>


RELEASES.md
Version 9.1-longOverhaul  (2023-02-25)
=========================
 * Fix firmware caused device error during startup
 * x50 (credits to Ryan & Darren):
   * Added software package 'ignore-ignition-line' for x50
   * Improved torque
   * Improved control stability at higher speeds
   * Reduced steering fighting torque
 * Perodua PSD:
   * Reduced steering fighting torque during manual override
   * Fix Alza/Veloz standstill brake and resume
   * Fix stock lane departure prevention and warning
   * Fix stock ADAS settings
   * Fix Alza/Veloz fingerprint clash with Ativa
   * Retuned lateral for better curve handling
   * Retuned longitudinal for Ativa and Alza
 * Perodua MG3 (credits to Kenneth):
   * More aggresive gas to close distance at higher speed


Version 9-longOverhaul  (20223-02-25)
=========================
 * Fix firmware caused device error during startup
 * x50 (credits to Ryan & Darren):
   * Added software package 'ignore-ignition-line' for x50
   * Improved torque
   * Improved control stability at higher speeds
   * Reduced steering fighting torque
 * Perodua PSD:
   * Reduced steering fighting torque during manual override
   * Fix Alza/Veloz standstill brake and resume
   * Fix stock lane departure prevention and warning
   * Fix stock ADAS settings
   * Fix Alza/Veloz fingerprint clash with Ativa
   * Retuned lateral for better curve handling
   * Retuned longitudinal for Ativa and Alza
 * Perodua MG3 (credits to Kenneth):
   * More aggresive gas to close distance at higher speed

Version 8-longOverhaul  (2022-11-30)
=========================
 * User configurable fan speed, power saver entry duration, vehicle skew and stopping distance
 * Perodua PSD: Fix brake bleeding
 * Perodua MG3: Fix distance and steering oscillation
 * Proton X50: Can disengage through cruise ready button
 * Proton X50: Hand touch warning if set to ICC instead of ACC
 * Added firmware recovery mechanism after a failed firmware flash
 * Support Perodua Alza (AV)
 * Support Toyota Veloz (AT)
 * Support Honda Civic 2022
 * Support Hyundai Ioniq HEV Plus 2017

Version 7-longOverhaul  (2022-10-25)
=========================
 * Recalculate MPC to allow harsher brakes for closer distance profiles
 * Added power saver toggle to shutdown device after 15 minutes of idle
 * Offset camera skew so that vehicles will slightly (negligible) lean to the right
 * Fix device re-registration problem after a reinstall
 * Toyota: Adjustable lead car follow profile
 * Perodua PSD: Stop-and-go function
 * Perodua PSD: Faster acceleration from standstill
 * Supported Proton X50 (Flagship)
 * Supported Honda City 2020 (V-Sensing)
 * Supported CRV 2020 (1.5 TCP, Black)
 * Supported Corolla Cross Hybrid

Version 6-longOverhaul  (2022-09-02)
=========================
 * New distance profile
 * Added more internal QC tools and vehicle porting tools

Version 5-longOverhaul  (2022-08-15)
=========================
 * bukapilot speed display matches stock speedometer
 * Cleaner settings UI
 * Rework of GPS time sync
 * New package feature thanks to @benmasato
 * Toyota: Enable stock Lane Departure Prevention when not engaged
 * Perodua Ativa & MFL: Adjustable lead car follow profile
 * From upstream openpilot v0.8.13:
   *  New driving model from upstream which is now trained on 1 million minutes of driving data
   *  Fixed lead training making lead predictions significantly more accurate
   *  Combined longitudinal planning now happens in a single MPC system
   *  New vision based forward collision warning
   *  New alert sounds
   *  New MPC acceleration lag compensation
   *  Fixed vehicleModelInvalid triggering due to false positives


Version 4-firstbatch  (2022-07-04)
=========================
 * Update UI: Training guide page, font change, add offroad alerts
 * Allow vision to leave power saver mode through ignition button and manual power on
 * Solve intermittent device power off
 * Change speed display to show GPS speed, NOT odometry speed
 * Toyota Alphard: Fix revving and wheelspeed scaling
 * Perodua PSD: Smoother low pump noise braking (lesser jerks)
 * Perodua PSD: Solve pump pressure bleeding issue
 * Perodua PSD: Enable stock Lane Departure Prevention when not engaged
 * Perodua Semi ACC: Add lateral tunes presets through settings
 * Support Toyota Corolla Cross 2021

Version 3-firstbatch  (2022-06-06)
=========================
 * Add full powersaver mode to 13 hours, battery can last at least 2 weeks idle with KommuAssist in power saver
 * Add stock ACC option in settings (only for Perodua)
 * Perodua PSD: Improve longitudinal
 * Perodua PSD: Add stock HUD warnings for front departure, forward collision warning & braking
 * Perodua PSD: Revert cruise speed set logic back to stock behaviour
 * Perodua PSD: Add brake pressure bleed warning
 * Perodua MG3: Fix odometer scaling
 * Increase standstill braking distance for all vehicles
 * QC: Add pre-fulfillment QC test
 * Fix minor UI display problems and release note popup
 * Add file corruption recovery mechanism during scons build

Version 2-firstbatch  (2022-05-11)
=========================
 * First bukapilot prebuilt release
 * Perodua: Remove horseriding
 * Perodua: 1.5x better longitudinal
 * Bug: Remove intermittent controls mismatch

Version 1-firstbatch  (2022-04-26)
=========================
 * Initial release of bukapilot based on openpilot v0.8.6
 * Supported Perodua Ativa 2020 - 2022 (AV)
 * Supported Perodua Myvi FL 2022 (AV)
 * Supported Toyota Alphard/Hybrid 2019-20
 * Supported Toyota Camry/Hybrid 2021-22
 * Supported Toyota Corolla Altis/Hybrid 2020-P
 * Supported Toyota Prius 2021-22
 * Supported Lexus ES/ES Hybrid 2019-21
 * Supported Lexus NX 2020
 * Supported Lexus UX/RX/Hybrid 2020-21
 * Supported Axia with KommuActuator2019 - 2022 (GXtra, Style, SE, AV)
 * Supported Bezza with KommuActuator 2020 - 2022 (X, AV)
 * Supported Myvi Gen 3 Pre-FL with KommuActuator 2017 - 2021 (G, X, H, AV)


selfdrive/updated.py
#!/usr/bin/env python3

# Safe Update: A simple service that waits for network access and tries to
# update every 10 minutes. It's intended to make the OP update process more
# robust against Git repository corruption. This service DOES NOT try to fix
# an already-corrupt BASEDIR Git repo, only prevent it from happening.
#
# During normal operation, both onroad and offroad, the update process makes
# no changes to the BASEDIR install of OP. All update attempts are performed
# in a disposable staging area provided by OverlayFS. It assumes the deleter
# process provides enough disk space to carry out the process.
#
# If an update succeeds, a flag is set, and the update is swapped in at the
# next reboot. If an update is interrupted or otherwise fails, the OverlayFS
# upper layer and metadata can be discarded before trying again.
#
# The swap on boot is triggered by launch_chffrplus.sh
# gated on the existence of $FINALIZED/.overlay_consistent and also the
# existence and mtime of $BASEDIR/.overlay_init.
#
# Other than build byproducts, BASEDIR should not be modified while this
# service is running. Developers modifying code directly in BASEDIR should
# disable this service.

import os
import datetime
import subprocess
import psutil
import shutil
import signal
import fcntl
import time
import threading
from pathlib import Path
from typing import List, Tuple, Optional

from common.basedir import BASEDIR
from common.markdown import parse_markdown
from common.params import Params
from selfdrive.hardware import EON, TICI, HARDWARE
from selfdrive.swaglog import cloudlog
from selfdrive.controls.lib.alertmanager import set_offroad_alert
from selfdrive.version import is_tested_branch

LOCK_FILE = os.getenv("UPDATER_LOCK_FILE", "/tmp/safe_staging_overlay.lock")
STAGING_ROOT = os.getenv("UPDATER_STAGING_ROOT", "/data/safe_staging")

NEOSUPDATE_DIR = os.getenv("UPDATER_NEOSUPDATE_DIR", "/data/neoupdate")

OVERLAY_UPPER = os.path.join(STAGING_ROOT, "upper")
OVERLAY_METADATA = os.path.join(STAGING_ROOT, "metadata")
OVERLAY_MERGED = os.path.join(STAGING_ROOT, "merged")
FINALIZED = os.path.join(STAGING_ROOT, "finalized")

DAYS_NO_CONNECTIVITY_MAX = 14     # do not allow to engage after this many days
DAYS_NO_CONNECTIVITY_PROMPT = 10  # send an offroad prompt after this many days

class WaitTimeHelper:
  def __init__(self, proc):
    self.proc = proc
    self.ready_event = threading.Event()
    self.shutdown = False
    signal.signal(signal.SIGTERM, self.graceful_shutdown)
    signal.signal(signal.SIGINT, self.graceful_shutdown)
    signal.signal(signal.SIGHUP, self.update_now)

  def graceful_shutdown(self, signum: int, frame) -> None:
    # umount -f doesn't appear effective in avoiding "device busy" on NEOS,
    # so don't actually die until the next convenient opportunity in main().
    cloudlog.info("caught SIGINT/SIGTERM, dismounting overlay at next opportunity")

    # forward the signal to all our child processes
    child_procs = self.proc.children(recursive=True)
    for p in child_procs:
      p.send_signal(signum)

    self.shutdown = True
    self.ready_event.set()

  def update_now(self, signum: int, frame) -> None:
    cloudlog.info("caught SIGHUP, running update check immediately")
    self.ready_event.set()

  def sleep(self, t: float) -> None:
    self.ready_event.wait(timeout=t)


def run(cmd: List[str], cwd: Optional[str] = None, low_priority: bool = False):
  if low_priority:
    cmd = ["nice", "-n", "19"] + cmd
  return subprocess.check_output(cmd, cwd=cwd, stderr=subprocess.STDOUT, encoding='utf8')


def set_consistent_flag(consistent: bool) -> None:
  os.sync()
  consistent_file = Path(os.path.join(FINALIZED, ".overlay_consistent"))
  if consistent:
    consistent_file.touch()
  elif not consistent:
    consistent_file.unlink(missing_ok=True)
  os.sync()


def set_params(new_version: bool, failed_count: int, exception: Optional[str]) -> None:
  params = Params()

  params.put("UpdateFailedCount", str(failed_count))

  last_update = datetime.datetime.utcnow()
  if failed_count == 0:
    t = last_update.isoformat()
    params.put("LastUpdateTime", t.encode('utf8'))
  else:
    try:
      t = params.get("LastUpdateTime", encoding='utf8')
      last_update = datetime.datetime.fromisoformat(t)
    except (TypeError, ValueError):
      pass

  if exception is None:
    params.delete("LastUpdateException")
  else:
    params.put("LastUpdateException", exception)

  # Write out release notes for new versions
  if new_version:
    try:
      with open(os.path.join(FINALIZED, "RELEASES.md"), "rb") as f:
        r = f.read().split(b'\n\n', 1)[0]  # Slice latest release notes
      try:
        params.put("ReleaseNotes", r.decode("utf-8"))
      except Exception:
        params.put("ReleaseNotes", r + b"\n")
    except Exception:
      params.put("ReleaseNotes", "")
  params.put_bool("UpdateAvailable", new_version)

  # Handle user prompt
  for alert in ("Offroad_UpdateFailed", "Offroad_ConnectivityNeeded", "Offroad_ConnectivityNeededPrompt"):
    set_offroad_alert(alert, False)

  now = datetime.datetime.utcnow()
  dt = now - last_update
  if failed_count > 15 and exception is not None:
    if is_tested_branch():
      extra_text = "Ensure the software is correctly installed"
    else:
      extra_text = exception
    set_offroad_alert("Offroad_UpdateFailed", True, extra_text=extra_text)
  elif dt.days > DAYS_NO_CONNECTIVITY_MAX and failed_count > 1:
    set_offroad_alert("Offroad_ConnectivityNeeded", True)
  elif dt.days > DAYS_NO_CONNECTIVITY_PROMPT:
    remaining = max(DAYS_NO_CONNECTIVITY_MAX - dt.days, 1)
    set_offroad_alert("Offroad_ConnectivityNeededPrompt", True, extra_text=f"{remaining} day{'' if remaining == 1 else 's'}.")


def setup_git_options(cwd: str) -> None:
  # We sync FS object atimes (which NEOS doesn't use) and mtimes, but ctimes
  # are outside user control. Make sure Git is set up to ignore system ctimes,
  # because they change when we make hard links during finalize. Otherwise,
  # there is a lot of unnecessary churn. This appears to be a common need on
  # OSX as well: https://www.git-tower.com/blog/make-git-rebase-safe-on-osx/

  # We are using copytree to copy the directory, which also changes
  # inode numbers. Ignore those changes too.
  git_cfg = [
    ("core.trustctime", "false"),
    ("core.checkStat", "minimal"),
  ]
  for option, value in git_cfg:
    run(["git", "config", option, value], cwd)


def dismount_overlay() -> None:
  if os.path.ismount(OVERLAY_MERGED):
    cloudlog.info("unmounting existing overlay")
    args = ["umount", "-l", OVERLAY_MERGED]
    if TICI:
      args = ["sudo"] + args
    run(args)


def init_overlay() -> None:

  overlay_init_file = Path(os.path.join(BASEDIR, ".overlay_init"))

  # Re-create the overlay if BASEDIR/.git has changed since we created the overlay
  if overlay_init_file.is_file():
    git_dir_path = os.path.join(BASEDIR, ".git")
    new_files = run(["find", git_dir_path, "-newer", str(overlay_init_file)])
    if not len(new_files.splitlines()):
      # A valid overlay already exists
      return
    else:
      cloudlog.info(".git directory changed, recreating overlay")

  cloudlog.info("preparing new safe staging area")

  params = Params()
  params.put_bool("UpdateAvailable", False)
  set_consistent_flag(False)
  dismount_overlay()
  if TICI:
    run(["sudo", "rm", "-rf", STAGING_ROOT])
  if os.path.isdir(STAGING_ROOT):
    shutil.rmtree(STAGING_ROOT)

  for dirname in [STAGING_ROOT, OVERLAY_UPPER, OVERLAY_METADATA, OVERLAY_MERGED]:
    os.mkdir(dirname, 0o755)

  if os.lstat(BASEDIR).st_dev != os.lstat(OVERLAY_MERGED).st_dev:
    raise RuntimeError("base and overlay merge directories are on different filesystems; not valid for overlay FS!")

  # Leave a timestamped canary in BASEDIR to check at startup. The device clock
  # should be correct by the time we get here. If the init file disappears, or
  # critical mtimes in BASEDIR are newer than .overlay_init, continue.sh can
  # assume that BASEDIR has used for local development or otherwise modified,
  # and skips the update activation attempt.
  consistent_file = Path(os.path.join(BASEDIR, ".overlay_consistent"))
  if consistent_file.is_file():
    consistent_file.unlink()
  overlay_init_file.touch()

  os.sync()
  overlay_opts = f"lowerdir={BASEDIR},upperdir={OVERLAY_UPPER},workdir={OVERLAY_METADATA}"

  mount_cmd = ["mount", "-t", "overlay", "-o", overlay_opts, "none", OVERLAY_MERGED]
  if TICI:
    run(["sudo"] + mount_cmd)
    run(["sudo", "chmod", "755", os.path.join(OVERLAY_METADATA, "work")])
  else:
    run(mount_cmd)

  git_diff = run(["git", "diff"], OVERLAY_MERGED, low_priority=True)
  params.put("GitDiff", git_diff)
  cloudlog.info(f"git diff output:\n{git_diff}")


def finalize_update() -> None:
  """Take the current OverlayFS merged view and finalize a copy outside of
  OverlayFS, ready to be swapped-in at BASEDIR. Copy using shutil.copytree"""

  # Remove the update ready flag and any old updates
  cloudlog.info("creating finalized version of the overlay")
  set_consistent_flag(False)

  # Copy the merged overlay view and set the update ready flag
  if os.path.exists(FINALIZED):
    shutil.rmtree(FINALIZED)
  shutil.copytree(OVERLAY_MERGED, FINALIZED, symlinks=True)

  run(["git", "reset", "--hard"], FINALIZED)
  run(["git", "submodule", "foreach", "--recursive", "git", "reset"], FINALIZED)

  set_consistent_flag(True)
  cloudlog.info("done finalizing overlay")


def handle_agnos_update(wait_helper: WaitTimeHelper) -> None:
  from selfdrive.hardware.tici.agnos import flash_agnos_update, get_target_slot_number

  cur_version = HARDWARE.get_os_version()
  updated_version = run(["bash", "-c", r"unset AGNOS_VERSION && source launch_env.sh && \
                          echo -n $AGNOS_VERSION"], OVERLAY_MERGED).strip()

  cloudlog.info(f"AGNOS version check: {cur_version} vs {updated_version}")
  if cur_version == updated_version:
    return

  # prevent an openpilot getting swapped in with a mismatched or partially downloaded agnos
  set_consistent_flag(False)

  cloudlog.info(f"Beginning background installation for AGNOS {updated_version}")
  set_offroad_alert("Offroad_NeosUpdate", True)

  manifest_path = os.path.join(OVERLAY_MERGED, "selfdrive/hardware/tici/agnos.json")
  target_slot_number = get_target_slot_number()
  flash_agnos_update(manifest_path, target_slot_number, cloudlog)
  set_offroad_alert("Offroad_NeosUpdate", False)


def handle_neos_update(wait_helper: WaitTimeHelper) -> None:
  from selfdrive.hardware.eon.neos import download_neos_update

  cur_neos = HARDWARE.get_os_version()
  updated_neos = run(["bash", "-c", r"unset REQUIRED_NEOS_VERSION && source launch_env.sh && \
                       echo -n $REQUIRED_NEOS_VERSION"], OVERLAY_MERGED).strip()

  cloudlog.info(f"NEOS version check: {cur_neos} vs {updated_neos}")
  if cur_neos == updated_neos:
    return

  cloudlog.info(f"Beginning background download for NEOS {updated_neos}")
  set_offroad_alert("Offroad_NeosUpdate", True)

  update_manifest = os.path.join(OVERLAY_MERGED, "selfdrive/hardware/eon/neos.json")

  neos_downloaded = False
  start_time = time.monotonic()
  # Try to download for one day
  while not neos_downloaded and not wait_helper.shutdown and \
        (time.monotonic() - start_time < 60*60*24):
    wait_helper.ready_event.clear()
    try:
      download_neos_update(update_manifest, cloudlog)
      neos_downloaded = True
    except Exception:
      cloudlog.info("NEOS background download failed, retrying")
      wait_helper.sleep(120)

  # If the download failed, we'll show the alert again when we retry
  set_offroad_alert("Offroad_NeosUpdate", False)
  if not neos_downloaded:
    raise Exception("Failed to download NEOS update")
  cloudlog.info(f"NEOS background download successful, took {time.monotonic() - start_time} seconds")


def check_git_fetch_result(fetch_txt: str) -> bool:
  err_msg = "Failed to add the host to the list of known hosts (/data/data/com.termux/files/home/.ssh/known_hosts).\n"
  return len(fetch_txt) > 0 and (fetch_txt != err_msg)


def check_for_update() -> Tuple[bool, bool]:
  setup_git_options(OVERLAY_MERGED)
  try:
    git_fetch_output = run(["git", "fetch", "--dry-run"], OVERLAY_MERGED, low_priority=True)
    return True, check_git_fetch_result(git_fetch_output)
  except subprocess.CalledProcessError:
    return False, False


def fetch_update(wait_helper: WaitTimeHelper) -> bool:
  cloudlog.info("attempting git fetch inside staging overlay")

  setup_git_options(OVERLAY_MERGED)

  git_fetch_output = run(["git", "fetch"], OVERLAY_MERGED, low_priority=True)
  cloudlog.info("git fetch success: %s", git_fetch_output)

  cur_hash = run(["git", "rev-parse", "HEAD"], OVERLAY_MERGED).rstrip()
  upstream_hash = run(["git", "rev-parse", "@{u}"], OVERLAY_MERGED).rstrip()
  new_version = cur_hash != upstream_hash
  git_fetch_result = check_git_fetch_result(git_fetch_output)

  cloudlog.info(f"comparing {cur_hash} to {upstream_hash}")
  if new_version or git_fetch_result:
    cloudlog.info("Running update")

    if new_version:
      cloudlog.info("git reset in progress")
      r = [
        run(["git", "reset", "--hard", "@{u}"], OVERLAY_MERGED, low_priority=True),
        run(["git", "clean", "-xdf"], OVERLAY_MERGED, low_priority=True ),
        run(["git", "submodule", "init"], OVERLAY_MERGED, low_priority=True),
        run(["git", "submodule", "update"], OVERLAY_MERGED, low_priority=True),
      ]
      cloudlog.info("git reset success: %s", '\n'.join(r))

      if EON:
        handle_neos_update(wait_helper)
      elif TICI:
        handle_agnos_update(wait_helper)

    # Create the finalized, ready-to-swap update
    finalize_update()
    cloudlog.info("openpilot update successful!")
  else:
    cloudlog.info("nothing new from git at this time")

  return new_version


def main() -> None:
  params = Params()

  if params.get_bool("DisableUpdates"):
    cloudlog.warning("updates are disabled by the DisableUpdates param")
    exit(0)

  ov_lock_fd = open(LOCK_FILE, 'w')
  try:
    fcntl.flock(ov_lock_fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
  except OSError as e:
    raise RuntimeError("couldn't get overlay lock; is another instance running?") from e

  # Set low io priority
  proc = psutil.Process()
  if psutil.LINUX:
    proc.ionice(psutil.IOPRIO_CLASS_BE, value=7)

  # Check if we just performed an update
  if Path(os.path.join(STAGING_ROOT, "old_openpilot")).is_dir():
    cloudlog.event("update installed")

  if not params.get("InstallDate"):
    t = datetime.datetime.utcnow().isoformat()
    params.put("InstallDate", t.encode('utf8'))

  overlay_init = Path(os.path.join(BASEDIR, ".overlay_init"))
  overlay_init.unlink(missing_ok=True)

  first_run = True
  last_fetch_time = 0.0
  update_failed_count = 0

  # Set initial params for offroad alerts
  set_params(False, 0, None)

  # Wait for IsOffroad to be set before our first update attempt
  wait_helper = WaitTimeHelper(proc)
  wait_helper.sleep(30)

  # Run the update loop
  #  * every 1m, do a lightweight internet/update check
  #  * every 10m, do a full git fetch
  while not wait_helper.shutdown:
    update_now = wait_helper.ready_event.is_set()
    wait_helper.ready_event.clear()

    # Don't run updater while onroad or if the time's wrong
    time_wrong = datetime.datetime.utcnow().year < 2019
    is_onroad = not params.get_bool("IsOffroad")
    if is_onroad or time_wrong:
      wait_helper.sleep(30)
      cloudlog.info("not running updater, not offroad")
      continue

    # Attempt an update
    exception = None
    new_version = False
    update_failed_count += 1
    try:
      init_overlay()

      internet_ok, update_available = check_for_update()
      if internet_ok and not update_available:
        update_failed_count = 0

      # Fetch updates at most every 10 minutes
      if internet_ok and (update_now or time.monotonic() - last_fetch_time > 60*10):
        new_version = fetch_update(wait_helper)
        update_failed_count = 0
        last_fetch_time = time.monotonic()

        if first_run and not new_version and os.path.isdir(NEOSUPDATE_DIR):
          shutil.rmtree(NEOSUPDATE_DIR)
        first_run = False
    except subprocess.CalledProcessError as e:
      cloudlog.event(
        "update process failed",
        cmd=e.cmd,
        output=e.output,
        returncode=e.returncode
      )
      exception = f"command failed: {e.cmd}\n{e.output}"
      overlay_init.unlink(missing_ok=True)
    except Exception as e:
      cloudlog.exception("uncaught updated exception, shouldn't happen")
      exception = str(e)
      overlay_init.unlink(missing_ok=True)

    try:
      set_params(new_version, update_failed_count, exception)
    except Exception:
      cloudlog.exception("uncaught updated exception while setting params, shouldn't happen")

    wait_helper.sleep(60)

  dismount_overlay()


if __name__ == "__main__":
  main()
